modules/internal/ExternalArray.chpl

```diff
@@ -96,34 +96,72 @@ module ExternalArray {
     return ret;
   }
 
+  proc convertStringOrBytes(ref arr: []): chpl_external_array
+    where arr.eltType == string || arr.eltType == bytes {
+
+    var wrapper: [arr.domain] chpl_byte_buffer;
+
+    // Move existing string buffers over to a new shell.
+    for i in 0..#arr.size {
+      ref item = arr[i];
+
+      // This call assumes ownership of the string's buffer.
+      var val = chpl__exportRetStringOrBytes(item);
+      wrapper[i] = val;
+    }
+
+    var ext = chpl_make_external_array_ptr_free(c_ptrTo(wrapper[0]),
+                                                wrapper.size: uint);
+
+    return ext;
+  }
+
   proc convertToExternalArray(in arr: []): chpl_external_array
-    where (getExternalArrayType(arr) == chpl_external_array) {
-    if (!isExternArrEltType(arr.eltType)) {
+    where getExternalArrayType(arr) == chpl_external_array {
+
+    if !isExternArrEltType(arr.eltType) {
       use HaltWrappers;
       safeCastCheckHalt("Cannot build an external array that stores " +
                         arr.eltType: string);
     }
-    if (!isIntegralType(arr.domain.idxType)) {
-      // Probably not reachable any more, but may become reachable again
-      // once support for interoperability with array types expands.
+
+    // Probably not reachable any more, but may become reachable again
+    // once support for interoperability with array types expands.
+    if !isIntegralType(arr.domain.idxType) then
       compilerError("cannot return an array with indices that are not " +
                     "integrals");
-    }
-    if (arr.domain.stridable) {
+
+    if arr.domain.stridable then
       compilerError("cannot return a strided array");
-    }
-    if (arr.domain.rank != 1) {
+    
+    if arr.domain.rank != 1 then
       compilerError("cannot return an array with rank != 1");
-    }
-    if (arr.domain.low != 0) {
+  
+    // 
+    // Can we just normalize the domain instead? Or would that be too
+    // unexpected?
+    //
+    if arr.domain.low != 0 then
       halt("cannot return an array when the lower bounds is not 0");
-    }
+
+    //
+    // TODO: If the array element type is string or bytes, then we need to
+    // silently convert it to an array of 'chpl_byte_buffer'. Then in
+    // compiler code, we should iterate through the elements and cast them
+    // to string/bytes as desired (in both cases creating a numpy array
+    // of Python objects).
```

Comment:
Is this still a TODO?

Reply:
Probably not a TODO, but I think it's good to allude to what the compiler is doing.